File: src/Features/Fileinfo_Proxy_Ips.php
<?php

namespace xltxlm\fileinfo\Features;

use xltxlm\redis\LockKey;

/**
 * 代理ip存在一个文件内,然后顺序读取出来作为代理,每次返回一个ip地址;
 */
class Fileinfo_Proxy_Ips extends Fileinfo_Proxy_Ips\Fileinfo_Proxy_Ips_implements
{
    private $cache_file_lines = [];

    public function __invoke(): string
    {
        if (empty($this->cache_file_lines)) {
            $ips = (new Fileinfo_Read_Lines($this->getFilepath()))
                ->__invoke();
            $proxyips = [];
            foreach ($ips as $ip) {
                if ($ip[0] == '#') {
                    continue;
                }
                $proxyips[] = $ip;
            }
            $this->cache_file_lines = $proxyips;
        }
        foreach ($this->cache_file_lines as $cache_file_line) {
            $LockKey = (new LockKey())
                ->setRedisConfig($this->getRedisCacheConfig())
                ->setKey("s{$ip}")
                ->setExpire($this->getProtect_Second())
                ->setValue($ip);
            if ($LockKey->__invoke() == false) {
                sleep($this->getProtect_Second());
            }
            return $cache_file_line;
        }
    }


}

File: go/Features/Fileinfo_Proxy_Ips.go
package Features

type Fileinfo_Proxy_Ips struct {
    /* ip超时锁,保护每个ip的请求频率 */
    RedisCacheConfig 

    /* ip保护的的间隔,默认1秒 */
    Protect_Second int

}

func NewFileinfo_Proxy_Ips(RedisCacheConfig ,Protect_Second int) *Fileinfo_Proxy_Ips{
    var this = new(Fileinfo_Proxy_Ips)
    this.SetRedisCacheConfig(RedisCacheConfig);
    this.SetProtect_Second(Protect_Second);
    return this
}

func (this *Fileinfo_Proxy_Ips)GetRedisCacheConfig() {
    return this.RedisCacheConfig;
}

func (this *Fileinfo_Proxy_Ips)SetRedisCacheConfig(RedisCacheConfig ) *Fileinfo_Proxy_Ips{
    this.RedisCacheConfig = RedisCacheConfig;
    return this
}
func (this *Fileinfo_Proxy_Ips)GetProtect_Second() int{
    return this.Protect_Second;
}

func (this *Fileinfo_Proxy_Ips)SetProtect_Second(Protect_Second int) *Fileinfo_Proxy_Ips{
    this.Protect_Second = Protect_Second;
    return this
}

/**
    */
func (this *Fileinfo_Proxy_Ips)__invoke()string{

}

